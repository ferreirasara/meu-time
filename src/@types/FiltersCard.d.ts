export type FiltersCardProps = {
  selectedCountry: string
  handleSelectCountry: (country: string) => void
  selectedLeague: number
  handleSelectLeague: (league: number) => void
  selectedSeason: number
  handleSelectSeason: (season: number) => void
  selectedTeam: number
  handleSelectTeam: (team: number) => void
}