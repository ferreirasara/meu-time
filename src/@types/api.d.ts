export type Team = {
  team: {
    id: number,
    name: string,
    code: string,
    country: string,
    founded: number,
    national: boolean,
    logo: string,
  },
  venue: {
    id: number,
    name: string,
    address: string,
    city: string,
    capacity: number,
    surface: string,
    image: string,
  },
}

export type Country = {
  name: string,
  code?: string,
  flag?: string,
}

export type League = {
  country: Country,
  league: {
    id: number,
    name: string,
    type: string,
    logo: string,
  }
  seasons: {
    year: number,
    start: string,
    end: string,
    current: boolean,
    coverage: {
      fixtures: {
        events: boolean,
        lineups: boolean,
        statistics_fixtures: boolean,
        statistics_players: boolean
      },
      standings: boolean,
      players: boolean,
      top_scorers: boolean,
      top_assists: boolean,
      top_cards: boolean,
      injuries: boolean,
      predictions: boolean,
      odds: boolean
    }
  }[],
}