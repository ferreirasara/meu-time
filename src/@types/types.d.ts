export type BaseData = {
  lastUpdated?: Date,
  countries?: Country[],
  leagues?: League[],
  seasons?: number[],
}

export type UserData = {
  email: string,
  firstname: string,
  lastname: string,
}

export type RequestData = {
  current: number,
  limitDay: number,
}

export type PlayersTableData = {
  id: number,
  name: string,
  age: number,
  nationality: string,
  photo: string,
}

export type PlayersTableProps = {
  selectedLeague: number
  selectedSeason: number
  selectedTeam: number
  handleError: (message: string) => void
}

export type GoalsChartProps = {
  data?: GeneralStatsData,
  loading: boolean,
}

export type GeneralStatsData = {
  totalDraws: number,
  totalPlayed: number,
  totalWins: number,
  totalLoses: number,
  mostUsedFormation: {
    formation: string,
    played: number,
  },
  goals: {
    against: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
    for: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
  }
}

export type GeneralStatsProps = {
  selectedLeague: number
  selectedSeason: number
  selectedTeam: number
  handleError: (message: string) => void
}

export type FiltersCardProps = {
  selectedCountry: string
  handleSelectCountry: (country: string) => void
  selectedLeague: number
  handleSelectLeague: (league: number) => void
  selectedSeason: number
  handleSelectSeason: (season: number) => void
  selectedTeam: number
  handleSelectTeam: (team: number) => void
  localStorageDataLoading: boolean
  handleError: (message: string) => void
}

export type Team = {
  team: {
    id: number,
    name: string,
    code: string,
    country: string,
    founded: number,
    national: boolean,
    logo: string,
  },
  venue: {
    id: number,
    name: string,
    address: string,
    city: string,
    capacity: number,
    surface: string,
    image: string,
  },
}

export type Country = {
  name: string,
  code?: string,
  flag?: string,
}

export type League = {
  country: Country,
  league: {
    id: number,
    name: string,
    type: string,
    logo: string,
  }
  seasons: {
    year: number,
    start: string,
    end: string,
    current: boolean,
    coverage: {
      fixtures: {
        events: boolean,
        lineups: boolean,
        statistics_fixtures: boolean,
        statistics_players: boolean
      },
      standings: boolean,
      players: boolean,
      top_scorers: boolean,
      top_assists: boolean,
      top_cards: boolean,
      injuries: boolean,
      predictions: boolean,
      odds: boolean
    }
  }[],
}

export type Player = {
  player: {
    id: number,
    name: string,
    firstname: string,
    lastname: string,
    age: number,
    birth: {
      date: string,
      place: string,
      country: string,
    },
    nationality: string,
    height: string,
    weight: string,
    injured: boolean,
    photo: string,
  },
  // statistics: {
  //   "team": {
  //     "id": 147,
  //     "name": "Coritiba",
  //     "logo": "https://media-1.api-sports.io/football/teams/147.png"
  //   },
  //   "league": {
  //     "id": 71,
  //     "name": "Serie A",
  //     "country": "Brazil",
  //     "logo": "https://media-2.api-sports.io/football/leagues/71.png",
  //     "flag": "https://media-3.api-sports.io/flags/br.svg",
  //     "season": 2016
  //   },
  //   "games": {
  //     "appearences": 6,
  //     "lineups": 6,
  //     "minutes": 438,
  //     "number": null,
  //     "position": "Defender",
  //     "rating": null,
  //     "captain": false
  //   },
  //   "substitutes": {
  //     "in": 0,
  //     "out": 3,
  //     "bench": 0
  //   },
  //   "shots": {
  //     "total": null,
  //     "on": null
  //   },
  //   "goals": {
  //     "total": 0,
  //     "conceded": null,
  //     "assists": null,
  //     "saves": null
  //   },
  //   "passes": {
  //     "total": null,
  //     "key": null,
  //     "accuracy": null
  //   },
  //   "tackles": {
  //     "total": null,
  //     "blocks": null,
  //     "interceptions": null
  //   },
  //   "duels": {
  //     "total": null,
  //     "won": null
  //   },
  //   "dribbles": {
  //     "attempts": null,
  //     "success": null,
  //     "past": null
  //   },
  //   "fouls": {
  //     "drawn": null,
  //     "committed": null
  //   },
  //   "cards": {
  //     "yellow": 0,
  //     "yellowred": 0,
  //     "red": 0
  //   },
  //   "penalty": {
  //     "won": null,
  //     "commited": null,
  //     "scored": null,
  //     "missed": null,
  //     "saved": null
  //   }
  // }[]
}

export type TeamStats = {
  lineups: {
    formation: string,
    played: number
  }[],
  fixtures: {
    draws: { home: number, away: number, total: number }
    loses: { home: number, away: number, total: number },
    played: { home: number, away: number, total: number },
    wins: { home: number, away: number, total: number },
  },
  goals: {
    against: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
    for: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
  }
}

export type InfoModalProps = {
  infoModalOpen: boolean
  onClose: () => void
}