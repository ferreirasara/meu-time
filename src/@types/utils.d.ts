export type BaseData = {
  lastUpdated?: Date,
  countries?: Country[],
  leagues?: League[],
  seasons?: number[],
}