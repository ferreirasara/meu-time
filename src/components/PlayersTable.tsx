import { Avatar, Divider, Table } from "antd"
import { ColumnsType } from "antd/es/table"
import { useEffect, useState } from "react"
import { getPlayersFromAPI } from "../utils/utils"
import { PlayersTableData, PlayersTableProps } from "../@types/types"

export const PlayersTable = ({
  selectedLeague,
  selectedSeason,
  selectedTeam,
  handleError,
}: PlayersTableProps) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [data, setData] = useState<PlayersTableData[]>([]);

  useEffect(() => {
    if (selectedLeague && selectedSeason && selectedTeam) {
      setLoading(true);
      getPlayersFromAPI(selectedLeague, selectedSeason, selectedTeam)
        .then(result => setData((result)?.map(player => ({
          id: player?.player?.id,
          name: player?.player?.name,
          age: player?.player?.age,
          nationality: player?.player?.nationality,
          photo: player?.player?.photo,
        }))))
        .catch(e => handleError(e?.toString()))
      setLoading(false);
    }
  }, [selectedLeague, selectedSeason, selectedTeam, handleError])

  const columns: ColumnsType<PlayersTableData> = [
    {
      dataIndex: 'photo',
      key: 'photo',
      render: (photo) => <Avatar size="small" src={photo} />
    },
    {
      title: 'Name',
      dataIndex: 'name',
      key: 'name',
    },
    {
      title: 'Idade',
      dataIndex: 'age',
      key: 'age',
    },
    {
      title: 'Nacionalidade',
      dataIndex: 'nationality',
      key: 'nationality',
    },
  ]

  return <div>
    <Divider>Jogadores</Divider>
    <Table
      loading={loading}
      columns={columns}
      dataSource={data}
      pagination={{ pageSize: 5 }}
      rowKey={'id'}
    />
  </div>
}