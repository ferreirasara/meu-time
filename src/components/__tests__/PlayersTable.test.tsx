import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { PlayersTable } from '../PlayersTable';
import { PlayersTableProps } from '../../@types/types';

const props: PlayersTableProps = {
  selectedLeague: 0,
  selectedSeason: 0,
  selectedTeam: 0,
  handleError: jest.fn(),
}

describe('PlayersTable', () => {
  beforeAll(() => {
    window.matchMedia = (query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })
  })

  it('match snapshot', () => {
    const login = renderer.create(<PlayersTable {...props} />).toJSON();
    expect(login).toMatchSnapshot();
  });

  it('renders table', () => {
    const { getByRole } = render(<PlayersTable {...props} />);

    const tableEl = getByRole("table");
    expect(tableEl).toBeInTheDocument();
  });

  it('renders all colunms', () => {
    const { getByText } = render(<PlayersTable {...props} />);

    const nameColunm = getByText(/Nome/);
    expect(nameColunm).toBeInTheDocument();

    const ageColunm = getByText(/Idade/);
    expect(ageColunm).toBeInTheDocument();

    const nationalityEl = getByText(/Nacionalidade/);
    expect(nationalityEl).toBeInTheDocument();
  });

  it('renders empty state', () => {
    const { getByText } = render(<PlayersTable {...props} />);

    const emptyEl = getByText(/No data/);
    expect(emptyEl).toBeInTheDocument();
  });
})