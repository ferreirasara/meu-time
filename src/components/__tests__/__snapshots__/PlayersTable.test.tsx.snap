// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayersTable match snapshot 1`] = `
<div>
  <div
    className="ant-divider css-dev-only-do-not-override-15rg2km ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center"
    role="separator"
  >
    <span
      className="ant-divider-inner-text"
      style={Object {}}
    >
      Jogadores
    </span>
  </div>
  <div
    className="ant-table-wrapper css-dev-only-do-not-override-15rg2km"
  >
    <div
      className="ant-spin-nested-loading css-dev-only-do-not-override-15rg2km"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-empty"
        >
          <div
            className="ant-table-container"
          >
            <div
              className="ant-table-content"
              onScroll={[Function]}
              style={Object {}}
            >
              <table
                style={
                  Object {
                    "tableLayout": "auto",
                  }
                }
              >
                <colgroup />
                <thead
                  className="ant-table-thead"
                >
                  <tr>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      scope={null}
                      style={Object {}}
                    />
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      scope="col"
                      style={Object {}}
                    >
                      Nome
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      scope="col"
                      style={Object {}}
                    >
                      Idade
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      scope="col"
                      style={Object {}}
                    >
                      Nacionalidade
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-placeholder"
                    style={
                      Object {
                        "display": null,
                      }
                    }
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={4}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <div
                        className="css-dev-only-do-not-override-15rg2km ant-empty ant-empty-normal"
                      >
                        <div
                          className="ant-empty-image"
                        >
                          <svg
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              fill="none"
                              fillRule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                cx="32"
                                cy="33"
                                fill="#f5f5f5"
                                rx="32"
                                ry="7"
                              />
                              <g
                                fillRule="nonzero"
                                stroke="#d9d9d9"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  fill="#fafafa"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          className="ant-empty-description"
                        >
                          No data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
